src/core_functions/scalar/math/numeric.cpp

```diff
@@ -1323,4 +1332,206 @@ ScalarFunctionSet LeastCommonMultipleFun::GetFunctions() {
 	return funcs;
 }
 
+//===--------------------------------------------------------------------===//
+// [divide_by_const]
+//===--------------------------------------------------------------------===//
+
+// This code is inspired by:
+// Lemire et al., Faster Remainder by Direct Computation: Applications to Compilers
+// and Software Libraries, Software: Practice and Experience 49 (6), 2019.
+// https://arxiv.org/abs/1902.01961
+
+static uhugeint_t ComputeM(uint64_t rhs) {
+	return uhugeint_t(0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF) / rhs + 1;
+}
+
+static uint64_t ComputeM(uint32_t rhs) {
+	return UINT64_C(0xFFFFFFFFFFFFFFFF) / rhs + 1;
+}
+
+static uint32_t ComputeM(uint16_t rhs) {
+	return UINT32_C(0xFFFFFFFF) / rhs + 1;
+}
+
+static uint16_t ComputeM(uint8_t rhs) {
+	return UINT32_C(0xFFFF) / rhs + 1;
+}
+
+static uint64_t ComputeM(int32_t rhs) {
+	if (rhs < 0) {
+		rhs = -rhs;
+	}
+	return UINT64_C(0xFFFFFFFFFFFFFFFF) / rhs + 1 + ((rhs & (rhs - 1)) == 0 ? 1 : 0);
+}
+
+static uint32_t ComputeM(int16_t rhs) {
+	if (rhs < 0) {
+		rhs = -rhs;
+	}
+	return UINT32_C(0xFFFFFFFF) / rhs + 1 + ((rhs & (rhs - 1)) == 0 ? 1 : 0);
+}
+
+static uint16_t ComputeM(int8_t rhs) {
+	if (rhs < 0) {
+		rhs = -rhs;
+	}
+	return UINT32_C(0xFFFF) / rhs + 1 + ((rhs & (rhs - 1)) == 0 ? 1 : 0);
+}
+
+// fastdiv computes (a / d) given precomputed M for d>1
+inline static uint64_t Fastdiv(uint64_t a, uhugeint_t m, uint64_t d) {
+	// Compute the bottom 32 bits, then the top 32 bits, add them, and only keep the top 32 bits
+	// The reason we need the bottom 32 bits, is because the carry will roll into the top 32 bits.
+	return ((((m & 0xFFFFFFFFFFFFFFFF) * a) >> 64) + ((m >> 64) * a)).upper;
+}
+
+// fastdiv computes (a / d) given precomputed M for d>1
+inline static uint32_t Fastdiv(uint32_t a, uint64_t m, uint32_t d) {
+	// Compute the bottom 32 bits, then the top 32 bits, add them, and only keep the top 32 bits
+	// The reason we need the bottom 32 bits, is because the carry will roll into the top 32 bits.
+	return ((((m & 0xFFFFFFFF) * a) >> 32) + ((m >> 32) * a)) >> 32;
+}
+
+inline static uint16_t Fastdiv(uint16_t a, uint32_t m, uint16_t d) {
+	// Compute the bottom 32 bits, then the top 32 bits, add them, and only keep the top 32 bits
+	// The reason we need the bottom 32 bits, is because the carry will roll into the top 32 bits.
+	return ((((m & 0xFFFF) * a) >> 16) + ((m >> 16) * a)) >> 16;
+}
+
+inline static uint8_t Fastdiv(uint8_t a, uint16_t m, uint8_t d) {
+	// Compute the bottom 32 bits, then the top 32 bits, add them, and only keep the top 32 bits
+	// The reason we need the bottom 32 bits, is because the carry will roll into the top 32 bits.
+	return ((((m & 0xFF) * a) >> 8) + ((m >> 8) * a)) >> 8;
+}
```

Comment:
I think we can also do more templating here:
```c++
template <class RESULT_TYPE, class OTHER_TYPE>
inline static uint8_t Fastdiv(RESULT_TYPE a, OTHER_TYPE m, RESULT_TYPE d) {
 	return ((((m & RESULT_TYPE(-1)) * a) >> sizeof(RESULT_TYPE) * 8) + ((m >> sizeof(RESULT_TYPE) * 8) * a)) >> sizeof(RESULT_TYPE) * 8;
 }
```
By templating you only have to write the code once, so there's less chance of making mistakes.

Templating usually works for all integral types except `hugeint` and `uhugeint`, as we've created these types ourselves.